mage_ai/streaming/sources/nats_js.py

```diff
@@ -0,0 +1,153 @@
+        self.thread.join()
+
+    async def ainit_client(self):
+        try:
+            self.nc = await nats.connect(
+                self.config.server_url,
+                error_cb=self.error_cb,
+                reconnected_cb=self.reconnected_cb,
+                disconnected_cb=self.disconnected_cb,
+                closed_cb=self.closed_cb,
+            )
+            self.js = self.nc.jetstream()
+
+            # Default consumer_name to stream_name if not provided
+            consumer_name = self.config.consumer_name or self.config.stream_name
+
+            await self.js.add_stream(name=self.config.stream_name, subjects=[self.config.subject])
+            self.psub = await self.js.pull_subscribe(self.config.subject, consumer_name)
+        except NoServersError as e:
+            self._print(f'Caught NoServersError while connecting to NATS server: {e}')
+        except Exception as e:
+            self._print(f'Caught exception while connecting to NATS server: {e}')
+
+    # Define callback methods
+    async def disconnected_cb(self):
+        self._print('Got disconnected!')
+
+    async def reconnected_cb(self):
+        self._print(f'Got reconnected to {self.nc.connected_url.netloc}')
+
+    async def error_cb(self, e):
+        self._print(f'There was an error: {e}')
+
+    async def closed_cb(self):
+        self._print('Connection is closed')
+
+    async def aclose_client(self):
+        await self.nc.close()
+
+    def init_client(self):
+        future = asyncio.run_coroutine_threadsafe(self.ainit_client(), self.loop)
+        future.result()
+
+    def close_client(self):
+        asyncio.run_coroutine_threadsafe(self.aclose_client(), self.loop)
+
+    def batch_read(self, handler: Callable):
+        self.init_client()
+
+        self._print("Starting batch read")
+
+        try:
+            while True:
+                messages = self.fetch_messages()
+                if not messages:
+                    break
+                handler(messages)  # Pass the entire batch to the handler
+        finally:
+            self.close_client()
+            self.stop_loop()
+
+    async def afetch_messages(self):
+        try:
+            msgs = await self.psub.fetch(self.config.batch_size, timeout=self.config.timeout)
+            return [json.loads(msg.data.decode()) for msg in msgs]
+        except nats.errors.TimeoutError:
+            return []
+
+    def fetch_messages(self):
+        future = asyncio.run_coroutine_threadsafe(self.afetch_messages(), self.loop)
+        return future.result()
+
+    def read(self, handler: Callable):
+        self.init_client()
+
+        self._print("Starting read")
+
+        try:
+            while True:
+                messages = self.fetch_messages()
+                if messages:
+                    for message in messages:
+                        handler(message)
+                else:
+                    time.sleep(1)  # Sleep for 1 second if no messages are available
```

Comment:
Given a timeout is supplied to the `psub.fetch` (which waits for at least one message to be returned), this should not be necessary to delay further.

Reply:
removed